x/staking/keeper/msg_server.go

```diff
@@ -383,17 +411,18 @@ func (k msgServer) Undelegate(goCtx context.Context, msg *types.MsgUndelegate) (
 		)
 	}
 
-	// tokenize share vs validator-bond delegation check if validator-bond delegation
-	validatorBondFactor := k.ValidatorBondFactor(ctx)
-	if delegation.ValidatorBond && !validatorBondFactor.IsNegative() {
-		maxTokenizeShareAfter := validator.TotalValidatorBondShares.Sub(shares).Mul(validatorBondFactor)
-		if maxTokenizeShareAfter.LT(validator.TotalLiquidShares) {
-			return nil, types.ErrInsufficientValidatorBondShares
+	// if this is a validator self-bond, the new liquid delegation cannot fall below the self-bond * bond factor
+	if delegation.ValidatorBond {
+		if err := k.SafelyDecreaseValidatorBond(ctx, validator, shares); err != nil {
+			return nil, err
 		}
+	}
 
-		// reduce total validator-bond delegation on unbond
-		validator.TotalValidatorBondShares = validator.TotalValidatorBondShares.Sub(shares)
-		k.SetValidator(ctx, validator)
+	// if this undelegation is from a liquid staking provider, the global and validator
+	// liquid counts should be decremented
+	if k.AccountIsLiquidStakingProvider(ctx, delegatorAddress) {
+		k.DecreaseTotalLiquidStakedTokens(ctx, tokens)
```

Comment:
Do you think we should emit some events when LS tokens are incremented/decremented, and when we perform other important LS related state changes?

Reply:
yeah that's a good idea - let's add them in a separate PR once this PR is merged